### PR TITLE
[WIN32SS][NTDDRAW][NTGDI] Implement some DxEng* callbacks

### PR DESCRIPTION
**Pull request by oleg-dubinskiy**
## Purpose

Implement some `DxEng`* functions required by MS DirectDraw stack (ddraw.dll & dxg.sys) (at least the system spams about them in the log, when it loads these components).
Call appropriate functions inside them, whose actually do needed things, and update their prototypes accordingly.

JIRA issue: [CORE-17561](https://jira.reactos.org/browse/CORE-17561)

## Proposed changes

Implement the following functions:

- `DxEngAltLockSurface`:
    - call `EngLockSurface` inside it (already implemented);
    - update the prototype.
- `DxEngDeleteSurface`:
    - call `EngDeleteSurface` inside it (already implemented);
    - update the prototype.
- `DxEngReferenceHdev`:
    - call `PDEVOBJ_vReference` inside it (already implemented).
- `DxEngSelectBitmap`:
    - call `NtGdiSelectBitmap` inside it (already implemented);
    - update the prototype.
- `DxEngSetBitmapOwner`:
    - call `GreSetBitmapOwner` inside it (already implemented);
    - update the prototype.
- `DxEngUnreferenceHdev`:
    - call `PDEVOBJ_vRelease` inside it (already implemented).
- Get rid of now unused `IntGdi`(`Un`)`referencePdev`. **My analysis proves that in Windows, `DxEngReferenceHdev` and `DxEngUnreferenceHdev` call `PDEVOBJ` functions directly. So our `IntGdi` ones are not needed anymore.**
- Add some `DPRINT`s.

Although it requires some more callbacks to be implemented, but unfortunately, for others, appropriate `NtGdi`* functions are not implemented yet. Moreover, they are absent at all. :frowning_face:

### CODE REVIEW

File: win32ss/gdi/eng/surface.c [hunk 1]
```diff
@@ -558,6 +558,16 @@ EngModifySurface(
...
+/*
+ * @implemented
+ */
+BOOL
+APIENTRY
+NtGdiEngDeleteSurface(
+    _In_ HSURF hsurf)
+{
+    return EngDeleteSurface(hsurf);
+}
```
Comment (GeoB99) [inline win32ss/gdi/eng/surface.c:570]: This doesn't look like a complete implementation. Whilst indeed the system call invokes `EngDeleteSurface` to free the GDI surface object, it un-secures the memory of that handle and decrements the reference count before that. As you can see in this image.
Capture

Although `MmUnsecureVirtualMemory` is not yet implemented in the Memory manager, you've to do some research to figure out how critical is the un-securing part. But invalidating the handle as-is can lead to abnormal behaviour.

**Comment by jimtabor** [comment]
Your imp is wrong!

This is how it is done in my Printer Driver support tree. Not yet committed.

__kernel_entry
BOOL
APIENTRY
NtGdiEngDeleteSurface(
    _In_ HSURF hsurf)
{
    PSURFACE psurf;

    psurf = SURFACE_ShareLockSurface(hsurf);
    if ( !psurf )
    {
        DPRINT1("Could not reference surface %p to delete\n", hsurf);
        return FALSE;
    }

    if ( !(psurf->flags & UMPD_SURFACE) )
    {
        SURFACE_ShareUnlockSurface(psurf);
        return FALSE;
    }

    GDIOBJ_vDeleteObject(&psurf->BaseObject);
    return TRUE;
}

**Comment by jimtabor** [comment]
It's pri use is called back from a user side print driver and passed to win32k via gdi32.

**Comment by jimtabor** [comment]
This is the way for this one too.

BOOL
APIENTRY
EngDeleteSurface(
    _In_ _Post_ptr_invalid_ HSURF hsurf)
{
    PSURFACE psurf;

    psurf = SURFACE_ShareLockSurface(hsurf);
    if (!psurf)
    {
        DPRINT1("Could not reference surface %p to delete\n", hsurf);
        return FALSE;
    }

    GDIOBJ_vDeleteObject(&psurf->BaseObject);
    return TRUE;
}

**Comment by jimtabor** [comment]
DxEngDeleteSurface should call EngDeleteSurface direct.

File: win32ss/gdi/eng/surface.c (hunk 1 above)
Comment (oleg-dubinskiy) [reply win32ss/gdi/eng/surface.c]: Not needed. Superseded by @jimtabor comment. :slightly_smiling_face:

**Comment by oleg-dubinskiy** [comment]
> DxEngDeleteSurface should call EngDeleteSurface direct.
> 
> NtGdiEngDeleteSurface is a call back from gdi32 for the Print Driver not wine hack DX code.

Thanks. Fixed now.
I guess if `NtGdiEngDeleteSurface` is not related to DX, it should be implemented in another PR.
Or, if it's already implemented in your branch, then I don't need to worry about it at all.

**Comment by jimtabor** [comment]
So far, these patches are valid.

**Comment by oleg-dubinskiy** [comment]
Sorry @jimtabor. :sweat_smile: I decided to improve my PR more (implemented more callbacks). Is it OK now on your opinion, with my new just pushed changes?

**Review by tkreuzer — APPROVED** [review]

**Review by binarymaster — APPROVED** [review]
LGTM.

- KVM: https://build.reactos.org/#builders/6/builds/31783
- VBox: https://build.reactos.org/#builders/3/builds/19262

**Pull request merged**